src/runtime_src/core/edge/drm/zocl/zocl_generic_cu.c

```diff
@@ -0,0 +1,222 @@
+	if (gcu->event > 0)
+		ret = POLLIN;
+
+	return ret;
+}
+
+static ssize_t generic_cu_read(struct file *filp, char __user *buf,
+			       size_t count, loff_t *ppos)
+{
+	struct generic_cu *gcu = filp->private_data;
+	DECLARE_WAITQUEUE(wait, current);
+	unsigned long flags;
+	ssize_t ret = 0;
+	s32 events = 0;
+
+	if (count != sizeof(s32))
+		return -EINVAL;
+
+	add_wait_queue(&gcu->waitq, &wait);
+
+	do {
+		set_current_state(TASK_INTERRUPTIBLE);
+
+		spin_lock_irqsave(&gcu->lock, flags);
+		if (gcu->event > 0) {
+			events = gcu->event;
+			gcu->event = 0;
+		}
+		spin_unlock_irqrestore(&gcu->lock, flags);
+
+		if (events > 0) {
+			if (copy_to_user(buf, &events, count))
+				ret = -EFAULT;
+			ret = sizeof(events);
+			break;
+		}
+
+		if (signal_pending(current)) {
+			ret = -ERESTARTSYS;
+			break;
+		}
+		schedule();
+	} while (1);
+
+	__set_current_state(TASK_RUNNING);
+	remove_wait_queue(&gcu->waitq, &wait);
+
+	return ret;
+}
+
+static ssize_t generic_cu_write(struct file *filp, const char __user *buf,
+				size_t count, loff_t *ppos)
+{
+	struct generic_cu *gcu = filp->private_data;
+	unsigned long flags;
+	s32 enable;
+
+	if (count != sizeof(s32))
+		return -EINVAL;
+
+	if (copy_from_user(&enable, buf, count))
+		return -EFAULT;
+
+	if (!gcu->info)
+		return -EINVAL;
+
+	spin_lock_irqsave(&gcu->lock, flags);
+	if (enable) {
+		if (__test_and_clear_bit(GCU_IRQ_DISABLED, &gcu->flag))
+			enable_irq(gcu->info->irq);
+	} else {
+		if (!__test_and_set_bit(GCU_IRQ_DISABLED, &gcu->flag))
+			disable_irq(gcu->info->irq);
+	}
+	spin_unlock_irqrestore(&gcu->lock, flags);
+
+	return sizeof(s32);
+}
+
+static const struct file_operations generic_cu_fops = {
+	.release	= generic_cu_release,
+	.poll		= generic_cu_poll,
+	.read		= generic_cu_read,
+	.write		= generic_cu_write,
+	.llseek		= noop_llseek,
+};
+
+static irqreturn_t generic_cu_isr(int irq, void *arg)
+{
+	struct generic_cu *gcu = arg;
+	unsigned long flags;
+
+
+	spin_lock_irqsave(&gcu->lock, flags);
+	gcu->event++;
+	if (!__test_and_set_bit(GCU_IRQ_DISABLED, &gcu->flag))
+		disable_irq_nosync(irq);
+	spin_unlock_irqrestore(&gcu->lock, flags);
+
+
+	wake_up_interruptible(&gcu->waitq);
+
+	return IRQ_HANDLED;
+}
+
+int _open_generic_cu(struct drm_zocl_dev *zdev, struct generic_cu_info *info)
+{
+	struct generic_cu *gcu;
+	int fd;
+	int ret;
+
+	gcu = kzalloc(sizeof(*gcu), GFP_KERNEL);
+	if (!gcu)
+		return -ENOMEM;
+
+	/* Do not use IRQF_SHARED! */
+	ret = request_irq(info->irq, generic_cu_isr, 0, info->name, gcu);
+	if (ret) {
+		DRM_INFO("%s: request_irq() failed\n", __func__);
+		return ret;
+	}
+
+	__set_bit(GCU_IRQ_DISABLED, &gcu->flag);
+	disable_irq(info->irq);
+
+	init_waitqueue_head(&gcu->waitq);
+	gcu->zdev = zdev;
+	gcu->info = info;
+	spin_lock_init(&gcu->lock);
+
+	fd = anon_inode_getfd("[generic_cu]", &generic_cu_fops, gcu, O_RDWR);
+	if (fd < 0)
+		kfree(gcu);
+
+	return fd;
+}
+
+int zocl_open_gcu(struct drm_zocl_dev *zdev, struct drm_zocl_ctx *ctx,
```

Comment:
if you're trying to implement a CU subdevice driver, passing in struct drm_zocl_dev seems to bring in unnecessary coupling b/w zocl and CU subdevice. let's chat.

Reply:
This is not a subdevice driver yet. But I would like generig_cu to be a subdevice driver in the future. This function is the only place that coupling zocl. We could easily change this later on.

I would add few callback functions in the generic_cu_info, so generic_cu could be the base driver for all other possible IPs and their variations.